lib/pbench/agent/tool_meister.py

```diff
@@ -1003,92 +1048,126 @@ def _send_client_status(self, status):
+
+    def init_tools(self, data: Dict[str, str]) -> int:
+        """Setup all registered persistent tools which have data collectors.
+
+        The Tool Data Sink will be setting up the actual processes which
+        collect data from these tools.
+
+        Arguments:
+
+            data: a dictionary of the arguments sent to the Tool Meister
+
+        Returns 0 on success, # of failures otherwise.
+        """
         try:
-            _tool_dir.mkdir()
-            # Remember this persistent tmp tool directory so that we can delete it
-            # when requested.
-            self.directories[data["directory"]] = (
-                _tool_dir if self._controller == self._hostname else _dir
-            )
+            _tmp_dir, _tool_dir = self._create_tool_directory(data["directory"])
         except Exception:
             self.logger.exception(
-                "Failed to create local result directory, %s", _tool_dir
+                "Failed to create local tool directory for %s", data["directory"]
             )
             self._send_client_status("internal-error")
-            return False
-        failures = 0
-        tool_cnt = 0
-        for name, tool_opts in sorted(self._tools.items()):
-            if name not in self.persistent_tool_names:
-                continue
-            tool_cnt += 1
-            tklass = self._tool_name_class_mappings[name]
-            try:
-                persistent_tool = tklass(
-                    name,
-                    tool_opts,
-                    pbench_install_dir=self.pbench_install_dir,
-                    tool_dir=_tool_dir,
-                    logger=self.logger,
-                )
-                persistent_tool.start()
-            except Exception:
-                self.logger.exception(
-                    "Failed to init PersistentTool %s running in background", name
-                )
-                failures += 1
-            else:
-                self._persistent_tools[name] = persistent_tool
-                self.logger.debug("NAME: " + name + "  TOOL OPTS: " + tool_opts)
+            return 1
+
+        # Remember this persistent tmp tool directory so that we can delete it
+        # when requested.
+        self.directories[data["directory"]] = _tmp_dir if _tmp_dir else _tool_dir
 
+        # Start all the persistent tools running.
+        started_tools = self._start_tools(self._persistent_tools, _tool_dir)
+
+        failures = len(self._persistent_tools) - len(started_tools)
         if failures > 0:
+            tool_cnt = len(self._persistent_tools)
             msg = f"{failures} of {tool_cnt} persistent tools failed to start"
             self._send_client_status(msg)
         else:
             self._send_client_status("success")
         return failures
 
-    def start_tools(self, data):
-        """start_tools - start all registered tools executing in the background
+    def start_tools(self, data: Dict[str, str]) -> int:
```

Comment:
That's intriguing: your code is still filled with `: dict` type hints, but here you went the distance for `Dict[str, str]`. I'd considered commenting on some of the simpler instances. I decided not to, as Python <3.9 ought to be able to handle the non-generic form `dict` successfully (as opposed to `dict[str,str]`)... and while `Dict[str,str]` is somewhat more expressive it still lacks any semantic information about the keys and values.

Even so, I prefer `Dict[str, str]`!

Reply:
Are type hints actually helping us?

So now at run time we have to import the typing module.  What is the purpose?

Are type hints really only for the human reader?

What I have read seems to say that is the case, and tools like `mypy` can help if applied, but do we leverage type hints?

I'd really like to see a material impact to type hints.

So far, I have only seen extra work applied to code with little to no upside (what bug has been caught by using type hints?  Has some tool helped us avoid a problem with type hints?).

I'll do as you ask, but we really need a clear benefit besides a beautiful soup full of UTF-8 glyphs added to the code.

Reply:
Yes, type hints are "mostly" for human readers, but I think they're extremely valuable in making massive chunks of code easier to digest. We could be using tools that would validate the parameters, and perhaps we should, but I think human readability is enough of a reason.

And while I agree that having to import `typing` is a pain, that'll be fixed in 3.9 where we can use builtin types as generics, like `dict[str, str]`; but when we get to drop 3.6 and such, I have no idea. (Sigh.)

Reply:
I have done my best to correct the use of `dict` and `Dict`.

Reply:
> type hints are "mostly" for human readers, but I think they're extremely valuable in making massive chunks of code easier to digest. We could be using tools that would validate the parameters, and perhaps we should, but I think human readability is enough of a reason.

Exactly this.  The type hints are only useful if we use them.  Once we have widespread use of them, then we can enable code inspection tools which rely on them.  (My IDE seems to pay attention to them in lots of cases.)

And, yes, I think that type hints have helped me find bugs during review.  (Presumably, if we were using inspection tools, the bug would have been found before I got there, but the hints helped either way.)

Mostly, I think they help with discipline in terms of writing and reading code:  they make us cognizant of what the types are supposed to be, even if Python isn't picky about that sort of thing -- and, my expectation is that that discipline will help us find bugs before runtime.